.travis.yml

```diff
@@ -1,43 +1,70 @@
-sudo: required
+sudo: false
 language: python
+python:
+- "3.5"
 addons:
+  apt:
+    packages:
+    - postgresql-contrib-9.5
   postgresql: "9.5"
 env:
   global:
-    - BOTO_CONFIG='' ETCDVERSION=2.2.5 ZKVERSION=3.4.6
+  - ETCDVERSION=2.3.1 ZKVERSION=3.4.6
   matrix:
-    - TEST_SUITE="python setup.py test"
-    - TEST_SUITE="behave"
-python:
-  - "2.7"
-  - "3.4"
-  - "3.5"
+  - TEST_SUITE="python setup.py"
+  - DCS="etcd" TEST_SUITE="behave"
+  - DCS="exhibitor" TEST_SUITE="behave"
+cache:
+  directories:
+  - $HOME/virtualenv/python2.7.9
+  - $HOME/virtualenv/python3.4.2
+  - $HOME/virtualenv/python3.5.0
 install:
   - |
+    set -e
+
     if [[ $TEST_SUITE == "behave" ]]; then
-        sudo bash -c '
-            /etc/init.d/postgresql stop
-            apt-get -y remove --purge postgresql-9.1 postgresql-9.2 postgresql-9.3 postgresql-9.4
-            apt-get -y autoremove
-            apt-key adv --keyserver keys.gnupg.net --recv-keys 7FCC7D46ACCC4CF8
-            echo "deb http://apt.postgresql.org/pub/repos/apt/ precise-pgdg main 9.5" >> /etc/apt/sources.list.d/postgresql.list
-            apt-get update
-            apt-get -y install postgresql-9.5
-            /etc/init.d/postgresql stop'
-        curl -L https://github.com/coreos/etcd/releases/download/v${ETCDVERSION}/etcd-v${ETCDVERSION}-linux-amd64.tar.gz | tar xz -C . --strip=1 --wildcards --no-anchored etcd
-        curl -L http://www.apache.org/dist/zookeeper/zookeeper-${ZKVERSION}/zookeeper-${ZKVERSION}.tar.gz | tar xz
-        mv zookeeper-${ZKVERSION}/conf/zoo_sample.cfg zookeeper-${ZKVERSION}/conf/zoo.cfg
-        zookeeper-${ZKVERSION}/bin/zkServer.sh start
-        while true; do
-            echo -e 'HTTP/1.0 200 OK\nContent-Type: application/json\n\n{"servers":["127.0.0.1"],"port":2181}' \
-              | nc -l 8181 &> /dev/null
-        done&
+        if [[ $DCS == "etcd" ]]; then
+            curl -L https://github.com/coreos/etcd/releases/download/v${ETCDVERSION}/etcd-v${ETCDVERSION}-linux-amd64.tar.gz | tar xz -C . --strip=1 --wildcards --no-anchored etcd
+        fi
+
+        if [[ $DCS == "exhibitor" ]]; then
+            curl -L http://www.apache.org/dist/zookeeper/zookeeper-${ZKVERSION}/zookeeper-${ZKVERSION}.tar.gz | tar xz
+            mv zookeeper-${ZKVERSION}/conf/zoo_sample.cfg zookeeper-${ZKVERSION}/conf/zoo.cfg
+            zookeeper-${ZKVERSION}/bin/zkServer.sh start
+            while true; do
+                echo -e 'HTTP/1.0 200 OK\nContent-Type: application/json\n\n{"servers":["127.0.0.1"],"port":2181}' | nc -l 8181 &> /dev/null
+            done&
+        fi
     fi
-  - pip install -r requirements.txt
-  - pip install behave codacy-coverage coverage coveralls
+
+    for pv in "2.7" "3.4" "3.5"; do
+        source ~/virtualenv/python${pv}/bin/activate
+        for p in '-r requirements.txt' 'behave codacy-coverage coverage coveralls flake8 mock pytest-cov pytest'; do
```

Comment:
Should we do 'test-requirements.txt' inside the patroni directory, listing all the test dependencies? Makes it easy to install all pre-requisites for running the test locally.